cli+log: more UX polishing around logging

knz opened the pull request:
This patch set makes minor changes that will enhance UX:

- it ensures that the warning about using an insecure cluster does show up on stderr even when logging (and stderr) is directed to files
- it also warns when --log-dir is not specified but there are multiple --stores, since this makes the determination of the log directory dependent on the order of the parameters (may be sensitive in case the parameter list is generated automatically)

knz committed:
cli: move the logging/profiling initialization to its own function.

knz committed:
log: add new Shout/Shoutf primitives that ensure log messages go to stderr.

knz committed:
cli+server: use log.Shout to ensure the user sees critical warnings/errors.

knz committed:
cli: avoid starting the log GC daemon if no log dir is specified.

knz committed:
cli: recommend using --log-dir if --store is specified multiple times

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14926)
<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but :-1: on the first commit.

---

Reviewed 4 of 4 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/build/info.go, line 67 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926#-KhgQP-avRGZgFl1scFi:-KhgQP-avRGZgFl1scFj:b-4m3o2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/371307f5d08ae305568c9553a90b977ced1d2daf/pkg/build/info.go#L67)):*
> ```Go
> }
> 
> var info = Info{
> ```

why is this worth the mutability?

---

*[pkg/cli/start.go, line 554 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926#-KhgRsReZOsOlNCXvE7I:-KhgRsReZOsOlNCXvE7J:b-t224pm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4768f39af3dd662764287188977fa46fd0a7ae11/pkg/cli/start.go#L554)):*
> ```Go
> 			}
> 			if newDir != "" {
> 				ambiguousLogDirs = true
> ```

break after this, and then you can just put the log here and remove `ambiguousLogDirs`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926#-:-KhgRRRFdPLrICjtBTw1:b-htgooj)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/build/info.go, line 67 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926#-KhgQP-avRGZgFl1scFi:-KhgSouX-yT-46UJwwbF:bmr9ryj) ([raw file](https://github.com/cockroachdb/cockroach/blob/371307f5d08ae305568c9553a90b977ced1d2daf/pkg/build/info.go#L67)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why is this worth the mutability?
</blockquote></details>

I figured since the value is always the same it could be computed just the one time. What's the disadvantage?

Anyway conceptually this specific change does not belong to this PR, so I split it away.

---

*[pkg/cli/start.go, line 554 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926#-KhgRsReZOsOlNCXvE7I:-KhgSvnMAuNjbeo93B52:b-kqs7i9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4768f39af3dd662764287188977fa46fd0a7ae11/pkg/cli/start.go#L554)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

break after this, and then you can just put the log here and remove `ambiguousLogDirs`
</blockquote></details>

ok for the early break, but the log must be performed after the call to Value.Set() below. Added a comment to clarify.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r6, 1 of 1 files at r7, 3 of 3 files at r8, 1 of 1 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14926)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYR
